### ISSUE TITLE
Implement Onboard keyboard ROS node

### ISSUE THREAD

**Issue by wzin** [EndPointCorp/lg_ros_nodes#127]
https://github.com/EndPointCorp/lg_chef/issues/463

tl;dr We need an [onboard keyboard](https://apps.ubuntu.com/cat/applications/precise/onboard/) to run anywhere. In most of the cases we will want to display onboard over a viewport that we mirror to the touchscreen and send touch to this onboard via touch routing ros node (EndpointCorp/lg_chef#474)

Requirements:
- make onboard react to` "activity_config" : "keyboard_enabled"` and prepare onboard for showing/hiding (if keyboard_enabled = true) and just ignore onboard messages if (keyboard_enabled = false)
- run one onboard instance per viewport (or be able to display onboard on any viewport) - let onboard have appropriate size depending on viewport size
- make all instances listen on common channel `/onboard/state`
- by design - onboard show/hide messages should be sent by client applications (client = application run inside of a container like google-chrome)

Dev notes:
- onboard config https://github.com/EndPointCorp/lg-live-image/blob/portal/config/includes.chroot/home/lg/etc/onboard.dconf
- @ezekiel says that you can set placement and size of onboard keyboard via dbus

**Comment by szymonlipinski** [comment]
Scripts for the current onboard ros node implementation: https://github.com/EndPointCorp/portal-ros/tree/development/catkin/src/onboard/scripts

Basically to show/hide onboard we need to send a dbus message: https://github.com/EndPointCorp/portal-ros/blob/development/catkin/src/onboard/scripts/listener.py#L20-L26

There is also a dbus function call which returns the onboard state

**Comment by wzin** [comment]
Ask @mvollrath to know how it's implemented in portalterminator.

**Comment by zdenekmaxa** [comment]
@mvollrath - could you please drop some references here pertinent to onboard implementation in portalterminator? thanks.

**Comment by mvollrath** [comment]
For terminator there were a few changes:

* A separate launcher node was factored out.
* Instead of keeping onboard running all the time, it is spun up on demand.
* The keyboard is coerced into the bottom of its viewport.

There is more to do to integrate it with mirroring.
* Need to run a node for each viewport.
* The node needs to be aware of the touch routing state so it can control visibility correctly.

In other words, each viewport must have a configured onboard node, but only the active touch viewport will show on a visibility state rising edge.

**Comment by wzin** [comment]
from converstaion with @mvollrath - onboard should follow touch router

**Comment by wzin** [comment]
Some notes from the meeting:

- tactile or any other rosbridge-connected webapp may publish `std_msgs/Bool` on `/onboard/visibility`

- develop onboard router ros node that will listen on `/onboard/visibility`
 - make it listen on director messages as well to decide on which viewport onboard should be spun on
 ```json
 {"activity_config": { 
 "route_touch": true,
"viewport": "your_viewport"
  }
 }
 ```
 - it needs to have a default viewport that it shows onboard on (in most cases `route_touch` won't be set
 - make onboard router publish list of show/hide events for "your_viewport" (or more viewports) on `/lg_onboard/activate`
 - run only one instance of lg_onboard router

- make an executor/actuator ros node that will subscribe to `lg_common/StringArray`:`/lg_onboard/activate` and will run on every viewport
 - make it show onboard on a viewport that it takes care of: start + applicationstate.VISIBLE with geometry

Scenario:
- user taps text field on the touchscreen
- the fields emits the message on `/onboard/activate` (it's in the kiosk extension)
- router remembers last director message - it has a default of `kiosk` but target viewport can be overriden with `route_touch`: true and `viewport` : `foobar`  in `activity_config` 
- router publishes `["kiosk"]` on `/lg_onboard/activate`
- kiosk lg_onboard executor spins onbaord and shows it on kiosk

Question:
- how do we show 3 onboards on a single displaynode on 3 separate viewports
(this may not be possible at all and we dont need it on portal but the answer may be important for later development)

**Comment by zdenekmaxa** [comment]
thanks @mvollrath for your yesterday's fixes - I confirm onboard shows and hides OK on portaldevcz. There is a problem with hiding when displayed on the wall, which I am looking into.
I have also continued with some refactoring and shortened things.
